sdks/java/io/solace/src/main/java/org/apache/beam/sdk/io/solace/read/UnboundedSolaceReader.java

```diff
@@ -133,14 +153,10 @@ public Instant getWatermark() {
 
   @Override
   public UnboundedSource.CheckpointMark getCheckpointMark() {
-    List<BytesXMLMessage> ackQueue = new ArrayList<>();
-    while (!elementsToCheckpoint.isEmpty()) {
-      BytesXMLMessage msg = elementsToCheckpoint.poll();
-      if (msg != null) {
-        ackQueue.add(msg);
-      }
-    }
-    return new SolaceCheckpointMark(active, ackQueue);
+    // It's possible for a checkpoint to be taken but never finalized.
+    // So we simply copy whatever safeToAckIds we currently have.
+    Map<Long, BytesXMLMessage> snapshotSafeToAckMessages = Maps.newHashMap(safeToAckMessages);
+    return new SolaceCheckpointMark(this::markAsAcked, active, snapshotSafeToAckMessages);
```

Comment:
The checkpoint mark does not need to be aware of whether the reader is or isn't active. The call to close may be observed by the finalizing thread to first close the underlying reader before atomically setting the active flag to false. Put a try/catch statement around the acknowledgement loop in the checkpoint mark's finalizer, if it trips on `IllegalStateException`, then the reader has been closed and none of the queued elements can be acknowledged.

Reply:
Good point, but I added the try catch only around the `ackMessage` method call. I think it could fail for other reasons, that a closed reader.